kamu-adapter-http/src/http_server_middleware.rs

```diff
@@ -132,25 +140,41 @@ where
 
             let dataset_repo = catalog.get_one::<dyn DatasetRepository>().unwrap();
 
-            let dataset = match dataset_repo.get_dataset(&dataset_ref).await {
-                Ok(ds) => ds,
-                Err(GetDatasetError::NotFound(err)) => {
-                    tracing::warn!("Dataset not found: {:?}", err);
-                    return Ok(Response::builder()
-                        .status(StatusCode::NOT_FOUND)
-                        .body(Default::default())
-                        .unwrap());
-                }
-                Err(err) => {
-                    tracing::error!("Could not get dataset: {:?}", err);
-                    return Ok(Response::builder()
-                        .status(StatusCode::INTERNAL_SERVER_ERROR)
-                        .body(Default::default())
-                        .unwrap());
-                }
-            };
+            if Self::is_allowed_to_create_dataset(&request) {
+                let dataset_builder = match dataset_repo.get_or_create_dataset(&dataset_ref).await {
+                    Ok(dsb) => Arc::new(dsb),
```

Comment:
Does putting a value into axum's extensions requires it to be `Clone`?
It's very unusual to see `Arc<Box<T>>` - it's a shared pointer that holds a unique pointer to T.

I guess my suggestion above would avoid this altogether.

Reply:
I think I had a weird error about Sync trait when passing the Box to lambda within the WS upgrade callback. But it seems it is solved now in this new form. Hard to tell.